src/language/patterns.md

```diff
@@ -0,0 +1,334 @@
+- [Collection literals][]
+
+This section describes common use cases for matching and destructuring with patterns, 
+combining various contexts, pattern types, and surrounding code constructs.
+
+### Variable declaration
+
+Anywhere Dart allows regular, single-name local variable declaration, you can also
+use a _pattern variable declaration_. If the pattern matches the value on the right
+of the declaration, it destructures the value, binding it to new local variables:
+
+```dart
+var (a, [b, c]) = ("str", [1, 2]);   // Declares new variables a, b, and c.
+```
+
+A pattern variable declaration must start with either `var` or `final`, followed
+by a pattern. 
+
+Use the [null-assert pattern][] to eliminate `null` in variable declarations.
+
+### Variable assignment 
+
+A _variable assignment pattern_ on the left side of an assignment destructures the
+matched object and assigns the values to _existing_ variables,
+instead of binding new ones. 
+
+One convenient use case is swapping the values of two variables without needing
+to declare a third temporary one:
+
+```dart
+var (a, b) = ('left', 'right');
+(b, a) = (a, b);                   // Swap
+print('$a $b');                    // Prints "right left".
+```
+
+### Destrcuturing multiple returns
+
+[Records][] allow aggregating and returning multiple values from a single function
+call. Patterns add the ability to destructure a record's fields
+directly into local variables, inline with the function call.
+
+Instead of individually declaring new local variables for each record field,
+like this:
+
+```dart
+main() {
+  var info = userInfo(json);
+  var name = info.$1;
+  var age = info.$2;
+}
+```
+You can call the function directly into a [variable declaration](#variable-declaration)
+or [assigment pattern](#variable-assignment),
+with a [record subpattern][record] and fields for each value in the return:
+
+```dart
+main() {
+  var (name, age) = userInfo(json);
+}
+```
+
+### Switch cases
+
+The `case` keyword is always followed by a pattern. _Case patterns_ are [refutable][];
+they allow control flow to either match and destructure the object being switched
+on, or continue execution if the object doesn't match. Case patterns can be any
+kind of pattern:
+
+- [**Constant patterns**][constant]: the object's value has to be equivalent to match.
+  ```dart
+  switch (number) {
+    case 1: ...          // Matches if number == 1  
+  }
+  ```
+- [**List**][list], [**map**][map], or [**record**][record] **patterns**: the elements are destructured if matched.
+  ```dart
+  switch (obj) {
+    case [a, b]: ...     // Matches if obj is a list with two elements  
+  }
+  ```  
+- [**Variable patterns**][variable]: the matched value binds to a new variable.
+  ```dart
+  switch ((1, 2)) {
+    case (var a, var b): ...    // Matches if 1 and 2 can be assigned to a and b
+  }
+  ```  
+- [**Logical**][logical] or [**relational**][relational] **patterns**: the object matches if the comparison to
```

Comment:
"The object matches if the comparison.." describes relational patterns only, so it reads a bit odd here when used to decribe "Logical or relational patterns".